### ISSUE TITLE
window.open doesn't support all standard functionality

### ISSUE THREAD

**Issue by bridiver** [atom/electron#4175]
We have made several changes to the window.open functionality to use the native blink/chromium code, but we are always working with embedded views so I'm looking for a little help to upstream this without breaking other use cases
Starting from https://github.com/brave/electron/commit/6ee28c3927a34495fb0a059ac3071e6e14358743 to https://github.com/brave/electron/commit/9dc98b1448f1efa1f2898fb714643df2b3474b12
We are working on getting the remaining electron tests passing and I'm happy to create a pull request as the basis for this, but some help/feedback would be welcome. Thanks!

**Comment by bridiver** [comment]
also have new tests that we can move from https://github.com/brave/browser-laptop to electron for the currently unsupported or incomplete functionality

**Comment by felixrieseberg** [comment]
This looks great - I assume you're talking about some of the `strWindowFeatures`? If so, I can really see how a bunch of other developers will rejoice if support is improved.

**Comment by bridiver** [comment]
for us it was mostly about complete cross-window scripting support with the correct security restrictions, but it comes with other benefits as well

**Comment by zcbenz** [comment]
I'm afraid we can not merge it.

Currently we have to guarantee that each WebContents maps to exact one renderer process, however your change would make the two WebContents share one renderer process.

We do this because:
* Many settings of `webPreferences` are process wide;
* Node is not perfectly safe when creating two Node environments in one process;
* Node leaks resources when a Node environment is destroyed (e.g. when a WebContents is destroyed);
* Chromium leaks memory when reloading/navigating/closing page;
* The renderer process has to be restarted when reloading/navigation happens, so we can make Node happy and assure not leaking resources.

Therefore as long as you are creating two WebContents in one renderer process, you are never going to fix all the failing specs of Electron. It also means we can never support `window.open` with normal ways.

I'm closing this as a duplicate of #1865.

**Comment by bridiver** [comment]
There is definitely another render process created when we open a child window using this code. When would they end up in the same process?

**Comment by zcbenz** [comment]
> There is definitely another render process created when we open a child window using this code. When would they end up in the same process?

It is probably because the URL passed to `window.open` is in a different origin, in this case Chromium would choose to create a new renderer process.

**Comment by bridiver** [comment]
The original issue we were targeting was Feedly login through Google. The child window is same origin and runs in it's own renderer process.

**Comment by bridiver** [comment]
We have tests for both same and different origin and both work correctly with correct origin-based access to parent/child

**Comment by zcbenz** [comment]
@bridiver Does the `contentWindow` property work? And can you print the `process.pid` in each window and see whether they are same.

**Comment by bridiver** [comment]
I'm not sure exactly what you are looking for with the contentWindow property. Can you give me an example you expect to fail?

**Comment by bridiver** [comment]
I did verify that process.pid is different in each and also by ps

**Comment by bridiver** [comment]
I can verify the postMessage, close, focus, blur and eval all work in both directions (eval only in same origin per http://www.w3.org/TR/html5/browsers.html#security-window)

**Comment by zcbenz** [comment]
@bridiver I have built your fork and opened the default app but following code doesn't seem to work in devtools console?

```js
var w = window.open('http://github.com', 'nodeIntegration=true');
```

**Comment by zcbenz** [comment]
> I can verify the postMessage, close, focus, blur and eval all work in both directions (eval only in same origin per http://www.w3.org/TR/html5/browsers.html#security-window)

Can you try whether the `contentWindow` property works?

**Comment by bridiver** [comment]
Blocked by popup blocking. Has to be user initiated

**Comment by bridiver** [comment]
contentWindow? From an iframe? Are you looking to see if you can establish bi-directional communication from different origins (postMessage)? Is there a test case in Electron I can run it against?

**Comment by zcbenz** [comment]
I created a local `main.html`:

```html
<html>
<head>
  <title>Electron</title>
</head>
<body>
<button onclick="doclick()">open</button>
<script type="text/javascript" charset="utf-8">
  function doclick() {
    console.log('click');
    window.w = window.open('http://github.com', 'nodeIntegration=true');
  }
</script>
</body>
</html>
```

Load it in a BrowserWindow, click the button, program crashed.

**Comment by bridiver** [comment]
Sorry, I should be clear that the current code only supports guest -> guest, but that is not a limitation of the code itself, that is just the only part we have implemented because that is all that we needed. Obviously we have to support all use cases before it can be upstreamed and that's what this issue was opened for. We are working on that portion of it, but any help is welcome. This code is working in our browser and we have several test cases for it there that we will move to Electron.

**Comment by bridiver** [comment]
Here's what I propose: we have a working implementation that I believe satisfies your requirements (separate renderer process/access to all properties by origin policy), but it needs to be expanded to cover window <-> window. Can we reopen this ticket pending those updates? If for some reason it can't cover all use cases we can implement it to only be used in the cases it does cover or only by flag.

**Comment by zcbenz** [comment]
This issue is closed because it is a duplicate of #1865.

I don't think we should ever add a flag that only solves one special use case while breaking all other apps, no one except you will use it and we are not able to maintain it.

**Comment by bridiver** [comment]
Fair enough, I think we can make it work for all cases anyway. I'll make a quick reference to our code in the other ticket in case anyone else wants to take a look.

**Comment by bridiver** [comment]
I have a few questions about your original comment so I can understand your concerns a little better and hopefully improve Electron for everyone:

1. Many settings of webPreferences are process wide
If the child window is in the same browsing context then shouldn't it have the same web preferences? If not it definitely wouldn't be in the same process and would have limited access to the child/opener

2. Node is not perfectly safe when creating two Node environments in one process;
If they are in the same process then wouldn't this only matter if node integration is enabled? 

3. Node leaks resources when a Node environment is destroyed (e.g. when a WebContents is destroyed);
The webContents live in the main process, so why would this be an issue for the renderer process?

4. The renderer process has to be restarted when reloading/navigation happens, so we can make Node happy and assure not leaking resources
I did notice that the child window loses the opener reference on reload (this doesn't happen in Chrome or Safari). This appears to be because webcontents.reload is actually implemented in Electron as a load of the current url. Either way wouldn't we just have to re-attach the webcontents to the renderer after restart?

Thanks!

**Comment by bridiver** [comment]
I made a mistake earlier, I was looking at the pid for the window, not the embedded webview. The two webviews do indeed share the same renderer process. I also see the memory issue you are referring to, but I have an almost complete fix for that. The issue is that a new node environment is created for every new script context in addition to the "global" environment that is created but never loaded. That isn't a problem right now because almost every navigation forks so you effectively end up with a process per page load. Destroying the node environment leaks memory, but a single node environment can have multiple contexts associated with it that can be sandboxed. My only remaining issue is the preload script, but I a proof of concept for that and just need to finish the implementation. The only time a node environment will be shared between more than one JS context is when they are connected through a scriptable opener. In cases of page navigation one associated JS context is swapped out for the next and different origin windows will be in different processes as normal.
